src/convert/fasta.rs

```diff
@@ -0,0 +1,379 @@
+                    sequence_starts.insert(name_buf.clone(), pos + 1);
+                    chr = name_buf.clone();
+                    name_buf.clear();
+                } else if pos > 0 {
+                    name_buf.push(*byte as char);
+                } else if *byte != b'>' {
+                    return Err(Error::FileFormat(
+                        "Fasta file does not start with a \">\" symbol".to_string(),
+                    ));
+                }
+                pos += 1;
+            }
+        }
+
+        //finish off last seq
+        if let Some(start) = soft_block_start {
+            let block = (start as usize)..in_seq_pos;
+            let soft_blocks_entry = soft_blocks.entry(chr.to_string()).or_insert_with(Vec::new);
+            soft_blocks_entry.push(block);
+        }
+        if let Some(start) = hard_block_start {
+            let block = (start as usize)..in_seq_pos;
+            let hard_blocks_entry = hard_blocks.entry(chr.to_string()).or_insert_with(Vec::new);
+            hard_blocks_entry.push(block);
+        }
+        sequence_lengths.push(SequenceLength::new(&chr, in_seq_pos));
+        Ok(Self {
+            file_path: None,
+            in_memory_data: None,
+            soft_blocks,
+            hard_blocks,
+            sequence_starts,
+            sequence_lengths,
+        })
+    }
+}
+
+/// Gives access to nucleotide sequences from a Fasta file
+struct FastaNucleotides {
+    reader: Box<dyn FileAccess>,
+    exhausted: bool,
+}
+
+impl Nucleotides for FastaNucleotides {
+    fn read_chunk(
+        &mut self,
+        buf: &mut [Nucleotide],
+    ) -> std::result::Result<Option<usize>, Box<dyn std::error::Error>> {
+        if self.exhausted {
+            Ok(None)
+        } else {
+            let mut newlines = 0;
+            let mut values = vec![0_u8; buf.len()];
+            let count = self.reader.read(&mut values)?;
+            if count == 0 {
+                self.exhausted = true;
+                return Ok(None);
+            }
+            for (i, byte) in values.iter().take(count).enumerate() {
+                if *byte == b'\n' {
+                    newlines += 1;
+                } else if *byte == b'>' {
+                    // we are running into the next sequence
+                    self.exhausted = true;
+                    return Ok(Some(i - newlines));
+                } else {
+                    let nuc: Nucleotide = (*byte).try_into()?;
+                    buf[i - newlines] = nuc;
+                }
+            }
+            Ok(Some(count - newlines))
+        }
+    }
+}
+
+impl<'a> SequenceRead<'a> for FastaReader {
+    fn sequence_lengths(&self) -> Cow<[SequenceLength]> {
+        Cow::Borrowed(&self.sequence_lengths)
+    }
+
+    fn nucleotides(&self, chr: &str) -> Box<dyn Nucleotides> {
+        let mut reader: Box<dyn FileAccess> = {
+            if let Some(file_path) = &self.file_path {
+                let file = File::open(file_path).unwrap();
```

Comment:
I still wonder if the `nucleotides` function should return a `Result` or not. And if it should, what would the error type be? Reminder: `nucleotides` is a trait method.

Reply:
Since this is a library, users are likely expecting to be able to handle errors. Returning `Result` thus makes sense unless you assume there is no same way to handle this error other than panic.